File: extra-addons/retana_home/models/home_dashboard.py
from odoo import models, fields, api
from odoo.exceptions import AccessError
import logging

_logger = logging.getLogger(__name__)

class HomeDashboard(models.Model):
    _name = 'home.dashboard'
    _description = 'Home Dashboard'
    
    name = fields.Char(string='Name', default='Home Dashboard')
    
    @api.model
    def get_installed_apps(self):
        """Get all installed applications, including Settings and Apps"""
        try:
            # Get all installed modules that are applications
            # COmentario de prueba
            apps = self.env['ir.module.module'].search([
                ('state', '=', 'installed'),
                ('application', '=', True)
            ])
            
            app_list = []
            for app in apps:
                if app.name == 'custom_home_dashboard':
                    # Skip the custom home dashboard module itself
                    continue
                
                # Get the menu associated with this app
                # First try by exact name match
                menu = self.env['ir.ui.menu'].search([
                    ('name', '=', app.shortdesc or app.name),
                    ('parent_id', '=', False)
                ], limit=1)
                
                if not menu:
                    # Try to find menu by similar name
                    menu = self.env['ir.ui.menu'].search([
                        ('name', 'ilike', app.shortdesc or app.name),
                        ('parent_id', '=', False)
                    ], limit=1)
                
                if not menu:
                    # Try to find by technical name patterns
                    menu = self.env['ir.ui.menu'].search([
                        ('name', 'ilike', app.name.replace('_', ' ')),
                        ('parent_id', '=', False)
                    ], limit=1)
                
                # If still no menu found, try to find any menu from this module
                if not menu:
                    # Search for menus defined in this module's XML files
                    menu_data = self.env['ir.model.data'].search([
                        ('module', '=', app.name),
                        ('model', '=', 'ir.ui.menu'),
                    ], limit=1)
                    if menu_data:
                        try:
                            potential_menu = self.env['ir.ui.menu'].browse(menu_data.res_id)
                            # Get the root menu (parent_id = False)
                            while potential_menu.parent_id:
                                potential_menu = potential_menu.parent_id
                            menu = potential_menu
                        except:
                            pass
                
                _logger.info(f"App: {app.shortdesc or app.name}, Menu found: {menu.id if menu else 'None'}")
                
                app_data = {
                    'id': app.id,
                    'name': app.shortdesc or app.name,
                    'technical_name': app.name,
                    'summary': app.summary or '',
                    'description': app.description or '',
                    'author': app.author or '',
                    'website': app.website or '',
                    'icon': self._get_app_icon(app.name),
                    'menu_id': menu.id if menu else False,
                    'web_icon': False,
                }
                app_list.append(app_data)
            
            # Add Settings manually
            app_list.append({
                'id': 'settings',
                'name': 'Settings',
                'technical_name': 'settings',
                'summary': 'System configuration and settings',
                'description': 'Access system settings and configurations',
                'author': 'Odoo',
                'website': '',
                'icon': 'fa-cogs',
                'menu_id': self.env.ref('base.menu_administration').id,
                'web_icon': False,
            })
            
            # CORRECCIÓN: Usar el menú correcto para Apps
            # Opción 1: Buscar por el external ID correcto
            try:
                apps_menu_id = self.env.ref('base.menu_management').id
            except:
                # Opción 2: Si no existe, buscar por varios nombres posibles
                apps_menu = self.env['ir.ui.menu'].search([
                    '|', '|', '|',
                    ('name', '=', 'Apps'),
                    ('name', '=', 'Applications'),
                    ('name', '=', 'Aplicaciones'),
                    ('name', 'ilike', 'app'),
                    ('parent_id', '=', False)
                ], limit=1)
                apps_menu_id = apps_menu.id if apps_menu else False
            
            app_list.append({
                'id': 'apps',
                'name': 'Apps',
                'technical_name': 'apps',
                'summary': 'Manage and install applications',
                'description': 'Access the app store to install and manage applications',
                'author': 'Odoo',
                'website': '',
                'icon': 'fa-th-large',
                'menu_id': apps_menu_id,
                'web_icon': False,
                # IMPORTANTE: Añadir acción específica para Apps
                'action_type': 'apps_store'
            })
            
            return app_list
            
        except AccessError:
            _logger.warning("Access denied when trying to get installed apps")
            return []
        except Exception as e:
            _logger.error("Error getting installed apps: %s", str(e))
            return []
    
    def _get_app_icon(self, module_name):
        """Get icon for the application"""
        # Common icons mapping
        icon_mapping = {
            'sale': 'fa-shopping-cart',
            'purchase': 'fa-shopping-bag',
            'account': 'fa-book',
            'stock': 'fa-truck',
            'hr': 'fa-users',
            'project': 'fa-tasks',
            'crm': 'fa-handshake-o',
            'website': 'fa-globe',
            'calendar': 'fa-calendar',
            'contacts': 'fa-address-book',
            'mail': 'fa-envelope',
            'documents': 'fa-folder',
            'manufacturing': 'fa-cogs',
            'quality': 'fa-check-circle',
            'maintenance': 'fa-wrench',
            'fleet': 'fa-car',
            'point_of_sale': 'fa-credit-card',
            'pos': 'fa-credit-card',
            'survey': 'fa-clipboard',
            'event': 'fa-ticket',
            'lunch': 'fa-cutlery',
            'expense': 'fa-money',
            'timesheet': 'fa-clock-o',
            'planning': 'fa-calendar-check-o',
            'social': 'fa-share-alt',
            'website_sale': 'fa-shopping-cart',
            'inventory': 'fa-cubes',
            'invoicing': 'fa-file-text',
            'marketing': 'fa-bullhorn',
            'helpdesk': 'fa-life-ring',
            'barcode': 'fa-barcode',
            'payment': 'fa-credit-card-alt',
        }
        
        # Try to find icon by module name
        for key, icon in icon_mapping.items():
            if key in module_name.lower():
                return icon
        
        # Default icon
        return 'fa-cube'
    
    @api.model
    def open_app(self, menu_id, action_type=None):
        """Open application by menu ID or special action"""
        
        # CORRECCIÓN: Manejar caso especial para Apps
        if action_type == 'apps_store':
            # Opción 1: Redirigir directamente a la acción de Apps
            return {
                'type': 'ir.actions.client',
                'tag': 'apps',
                'target': 'current',
            }
        
        # ALTERNATIVA: Si la opción anterior no funciona, usar esta
        # if action_type == 'apps_store':
        #     return {
        #         'type': 'ir.actions.act_window',
        #         'name': 'Apps',
        #         'res_model': 'ir.module.module',
        #         'view_mode': 'kanban,tree,form',
        #         'domain': [('application', '=', True)],
        #         'context': {'search_default_name': 1},
        #         'target': 'current',
        #     }
        
        if not menu_id:
            return False
            
        return {
            'type': 'ir.actions.client',
            'tag': 'menu',
            'params': {'menu_id': menu_id}
        }

File: extra-addons/retana_home/static/src/js/home_dashboard.js
/** @odoo-module **/

import { registry } from "@web/core/registry";
import { Component, useState, onMounted } from "@odoo/owl";
import { useService } from "@web/core/utils/hooks";

class HomeDashboard extends Component {
    setup() {
        this.orm = useService("orm");
        this.actionService = useService("action");
        this.state = useState({
            apps: [],
            loading: true,
            error: false
        });
        
        onMounted(() => {
            this.loadApps();
        });
    }

    async loadApps() {
        try {
            this.state.loading = true;
            this.state.error = false;
            
            const apps = await this.orm.call(
                'home.dashboard',
                'get_installed_apps',
                []
            );
            
            console.log('Loaded apps:', apps);
            this.state.apps = apps;
            this.state.loading = false;
            
        } catch (error) {
            console.error('Error loading apps:', error);
            this.state.error = true;
            this.state.loading = false;
        }
    }

    async openApp(app) {
        console.log('=== Opening app ===');
        console.log('App data:', app);
        console.log('Menu ID:', app.menu_id);
        console.log('Action type:', app.action_type);
        
        try {
            // Special handling for Apps menu
            if (app.action_type === 'apps_store') {
                console.log('Opening apps store...');
                await this.actionService.doAction('base.open_module_tree');
                return;
            }
            
            if (app.menu_id) {
                console.log('Attempting to open menu with ID:', app.menu_id);
                
                // Get the action associated with this menu
                const menuAction = await this.orm.call(
                    'ir.ui.menu',
                    'read',
                    [[app.menu_id], ['action']]
                );
                
                console.log('Menu action:', menuAction);
                
                if (menuAction && menuAction[0] && menuAction[0].action) {
                    const actionString = menuAction[0].action;
                    const actionId = parseInt(actionString.split(',')[1]);
                    console.log('Executing action ID:', actionId);
                    await this.actionService.doAction(actionId);
                } else {
                    // Try direct URL navigation
                    console.log('No action found, using URL navigation');
                    window.location.href = `/web#menu_id=${app.menu_id}`;
                }
            } else {
                // Fallback: Search for the app's main action
                console.warn('No menu found for app:', app.name);
                await this.openAppBySearch(app);
            }
        } catch (error) {
            console.error('Error opening app:', error);
            console.error('Error details:', error.message, error.stack);
            // Last resort: try to navigate using window location
            this.openAppByUrl(app);
        }
    }

    async openAppBySearch(app) {
        try {
            // Try to find and execute the app's main action
            const actions = await this.orm.searchRead(
                'ir.actions.act_window',
                [['name', 'ilike', app.name]],
                ['id', 'name', 'res_model', 'view_mode'],
                { limit: 1 }
            );
            
            if (actions.length > 0) {
                await this.actionService.doAction(actions[0].id);
            }
        } catch (error) {
            console.error('Error in openAppBySearch:', error);
        }
    }

    openAppByUrl(app) {
        // Last resort: navigate using URL
        if (app.menu_id) {
            window.location.href = `/web#menu_id=${app.menu_id}`;
        }
    }

    getAppIcon(app) {
        // If the app has a web_icon, use it
        if (app.web_icon) {
            return app.web_icon;
        }
        
        // Otherwise use the FontAwesome icon
        return `fa ${app.icon}`;
    }
}

HomeDashboard.template = "custom_home_dashboard.HomeDashboard";

registry.category("actions").add("home_dashboard", HomeDashboard);